exonum-light-client/src/main/java/com/exonum/client/ExonumHttpClient.java

```diff
@@ -141,19 +149,76 @@ public BlockResponse getBlockByHeight(long height) {
   }
 
   @Override
-  public BlocksResponse getBlocks(int count, BlockFilteringOption blockFilter, long heightMax,
+  public BlocksRange getBlocks(long fromHeight, long toHeight, BlockFilteringOption blockFilter,
       BlockTimeOption timeOption) {
-    checkArgument(0 < count,
-        "Requested number of blocks should be positive number but was %s", count);
-    return doGetBlocks(count, blockFilter, heightMax, timeOption);
+    checkArgument(0 <= fromHeight, "First block height (%s) must be non-negative", fromHeight);
+    checkArgument(fromHeight <= toHeight,
+        "First block height (%s) should be less than or equal to the last block height (%s)",
+        fromHeight, toHeight);
+
+    // 'maximum' as when skipping empty the actual might be way smaller
+    int maxSize = Math.toIntExact(toHeight - fromHeight + 1);
+    List<Block> blocks = new ArrayList<>(maxSize);
+    for (long rangeLast = toHeight; rangeLast >= fromHeight; ) {
+      int remainingBlocks = Math.toIntExact(rangeLast - fromHeight + 1);
+      int numBlocks = min(remainingBlocks, MAX_BLOCKS_PER_REQUEST);
+      BlocksResponse blocksResponse = doGetBlocks(numBlocks, blockFilter, rangeLast, timeOption);
+
+      blocks.addAll(blocksResponse.getBlocks());
+
+      rangeLast = blocksResponse.getBlocksRangeStart() - 1;
+    }
+
+    return postProcessResponseBlocks(fromHeight, toHeight, blocks);
   }
 
   @Override
-  public BlocksResponse getLastBlocks(int count, BlockFilteringOption blockFilter,
+  public BlocksRange getLastBlocks(int count, BlockFilteringOption blockFilter,
       BlockTimeOption timeOption) {
     checkArgument(0 < count,
         "Requested number of blocks should be positive number but was %s", count);
-    return doGetBlocks(count, blockFilter, null, timeOption);
+
+    List<Block> blocks = new ArrayList<>(count);
+    // The first request does not specify the maximum height to get the top blocks
+    long blockchainHeight = Long.MIN_VALUE;
+    Long nextHeight = null;
+    int remainingBlocks = count;
+    while (remainingBlocks > 0
+        && (nextHeight == null || nextHeight >= GENESIS_BLOCK_HEIGHT)) {
+      int numBlocks = min(remainingBlocks, MAX_BLOCKS_PER_REQUEST);
+      BlocksResponse blocksResponse = doGetBlocks(numBlocks, blockFilter, nextHeight, timeOption);
+
+      blocks.addAll(blocksResponse.getBlocks());
+
+      nextHeight = blocksResponse.getBlocksRangeStart() - 1;
+      blockchainHeight = max(blockchainHeight, blocksResponse.getBlocksRangeEnd() - 1);
+      remainingBlocks = Math.toIntExact(count - (blockchainHeight - nextHeight));
```

Comment:
looks like we can remove `blockchainHeight` and use `blocks.size()` instead

Reply:
No, we can't, because (a) blockchainHeight is not known until the first request completes; (b) it is needed to infer the number of remaining blocks and/or fromHeight.